README.md
# number-to-words
A library to convert a number into words.

![Clojars Version](https://img.shields.io/clojars/v/io.github.yuhribernardes/number-to-words?color=%2336B8DD&logo=Clojure&logoColor=white&style=for-the-badge)

## Installation

```edn
[io.github.yuhribernardes/number-to-words "0.2.0"]
```

## Usage

### Basic usage

``` clojure
(ns my-ns
 (:require [number-to-words.parse :as ntw]
           [number-to-words.client :as ntw-cli]))
 
(def client (ntw-cli/create :en)) ;; create a client for the desired language

(ntw/number->words client 123) ;;=> "one hundred and twenty three"
```

### Extending your own language

``` clojure
(ns my-ns
 (:require [number-to-words.parse :as ntw]
           [number-to-words.client :as ntw-cli]))

(def my-specs (ntw-cli/assoc-lang
               ntw-cli/specs
               :pt
               {:zero-case "zero"
                :natural {1 "um"
                          2 "dois"
                          3 "três"
                          4 "quatro"}
                :pows ["milhão" "mil" nil]}))

(def my-client (ntw-cli/create my-specs :pt))

(ntw/number->words my-client 0) ;;=> "zero"
(ntw/number->words my-client 1) ;;=> "um"
(ntw/number->words my-client 2) ;;=> "dois"
(ntw/number->words my-client 3) ;;=> "três"
(ntw/number->words my-client 4) ;;=> "quatro"
```

## Supported language
- English (`:en`)

## Deploy

1. Configure auth envs

``` sh
export CLOJARS_USER=<your-clojars-username>
export CLOJARS_PWD=<your-deploy-token>
```

> Notice you need to generate deploy tokens at your clojars account

2. Deploy it!!

``` sh
lein deploy release
```

> You can also see deploy options [here][clojars-deploy]

## Tests

### Standalone

``` sh
lein kaocha
```

### For development
It's highly recommended to run tests in watch mode to see what is happening while making changes in the code

``` sh
lein kaocha --watch
```

## Development environment

### Default profiles

You need to configure your repl to start with te profiles described in the list below:
- kaocha

If you use Emacs, it's highly recommended to create a `dir-locals.el` file with the given content below:

``` emacs-lisp
((clojure-mode
  (cider-lein-global-options . "with-profile +kaocha")))
```

## To Do

- [ ] Auto generate library version based on [Semantic Versioning][semver]
- [ ] Better namespace organizations (move core functions to their own namespaces)
- [ ] Add better Clojure Docs
- [x] Configure clj-kondo to not throw errors to some forms (e.g. nubank matcher-combinators `match?`)

## License

Copyright © 2022

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.

[semver]: https://semver.org
[clojars-deploy]: https://github.com/technomancy/leiningen/blob/master/doc/DEPLOY.md#deploying-libraries


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## 1.0.0 - 2022-03-03
- [BREAKING] Now the parser is a client based operation.
- Clients are extensibles based on a language. You can create your own language specifications
- Doc update

## 0.2.2 - 2022-02-21
- Configured clojars deploy
- Added first README doc
- Added clj docs to core function

## 0.2.0 - 2022-02-21
- Added number to word convertion

## 0.1.0 - 2022-02-14
### Added
- Files from the new template.


test/number_to_words/core_test.clj
(ns number-to-words.core-test
  (:require [clojure.test :refer :all]
            [number-to-words.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


test/number_to_words/parse_test.clj
(ns number-to-words.parse-test
  (:require [number-to-words.parse :as parse]
            [number-to-words.client :as n-client]
            [matcher-combinators.test]
            [matcher-combinators.matchers :as m]
            [clojure.test :refer [deftest testing is are]]))

(defn create-case [in out]
  {:in in :out out})

(deftest number->digit-groups
  (testing "should split number into trios of digit"
    (are [in out]
        (match? out (parse/number->digit-groups in))

      1 (list (list 0 0 0) (list 0 0 0) (list 0 0 1))
      11 (list (list 0 0 0) (list 0 0 0) (list 0 1 1))
      111 (list (list 0 0 0) (list 0 0 0) (list 1 1 1))
      111111 (list (list 0 0 0) (list 1 1 1) (list 1 1 1))
      111111111 (list (list 1 1 1) (list 1 1 1) (list 1 1 1)))))

(deftest get-after-hundreds
  (let [client (n-client/create :en)]
    (testing "[en] should get words for natural tens and units for"
      (are [group out]
          (match? out (parse/get-after-hundreds client group))
       [0 1] "one"
       [0 2] "two"
       [0 3] "three"
       [0 4] "four"
       [0 5] "five"
       [0 6] "six"
       [0 7] "seven"
       [0 8] "eight"
       [0 9] "nine"
       [1 0] "ten"
       [1 1] "eleven"
       [1 2] "twelve"
       [1 3] "thirteen"
       [1 4] "fourteen"
       [1 5] "fifteen"
       [1 6] "sixteen"
       [1 7] "seventeen"
       [1 8] "eighteen"
       [1 9] "nineteen"
       [2 0] "twenty"
       [3 0] "thirty"
       [4 0] "forty"
       [5 0] "fifty"
       [6 0] "sixty"
       [7 0] "seventy"
       [8 0] "eighty"
       [9 0] "ninety")

      (testing "[en] should get words for non natural tens and units groups"
        (are [group out]
            (match? out (parse/get-after-hundreds client group))
         [2 1] "twenty one"
         [3 2] "thirty two"
         [4 3] "forty three"
         [5 4] "fifty four"
         [6 5] "sixty five"
         [7 6] "seventy six"
         [8 7] "eighty seven"
         [9 8] "ninety eight"
         [9 9] "ninety nine"))

      (testing "[en] shouldn't get any words when there's only zeros"
        (is (match? nil
                    (parse/get-after-hundreds client [0 0])))))))

(deftest get-hundreds
  (let [client (n-client/create :en)]
    (testing "Should return the name without 'and' complement"
      (are [group out]
          (match? out (parse/get-hundreds client group))

        [1 0 0] "one hundred"
        [2 0 0] "two hundred"
        [3 0 0] "three hundred"
        [4 0 0] "four hundred"
        [5 0 0] "five hundred"
        [6 0 0] "six hundred"
        [7 0 0] "seven hundred"
        [8 0 0] "eight hundred"
        [9 0 0] "nine hundred")

      (testing "Should return the word with 'and' complement"
        (are [group out]
            (match? out (parse/get-hundreds client group))
         [1 0 1] "one hundred and"
         [2 1 0] "two hundred and"
         [3 1 1] "three hundred and"
         [4 1 0] "four hundred and"
         [5 0 1] "five hundred and"
         [6 1 0] "six hundred and"
         [7 1 1] "seven hundred and"
         [8 1 1] "eight hundred and"
         [9 0 1] "nine hundred and")))))

(deftest group->word
  (let [client (n-client/create :en)]
    (testing "Should return the words with power specs"
      (are [power group out]
          (match? out (parse/group->words client power group))

        0 [0 0 1] "one million"
        1 [0 0 1] "one thousand"
        1 [0 1 1] "eleven thousand"
        2 [0 0 1] "one"
        0 [1 3 8] "one hundred and thirty eight million"
        1 [7 1 9] "seven hundred and nineteen thousand"
        2 [3 4 7] "three hundred and forty seven"))

    (testing "Should return nil when every number is zero"
      (let [test-cases [(create-case [0 [0 0 0]] nil)
                        (create-case [1 [0 0 0]] nil)
                        (create-case [2 [0 0 0]] nil)]]
        (are [power group]
            (match? nil (parse/group->words client power group))
          0 [0 0 0]
          1 [0 0 0]
          2 [0 0 0])))))

(deftest number->words
  (let [client (n-client/create :en)]
   (testing "should return the correct words for the given number"
     (are [number words]
         (match? words (parse/number->words client number))

       1 "one"
       2 "two"
       3 "three"
       4 "four"
       5 "five"
       6 "six"
       7 "seven"
       8 "eight"
       9 "nine"
       10 "ten"
       11 "eleven"
       12 "twelve"
       13 "thirteen"
       14 "fourteen"
       15 "fifteen"
       16 "sixteen"
       17 "seventeen"
       18 "eighteen"
       19 "nineteen"
       20 "twenty"
       30 "thirty"
       40 "forty"
       50 "fifty"
       60 "sixty"
       70 "seventy"
       80 "eighty"
       90 "ninety"
       173 "one hundred and seventy three"
       10101 "ten thousand and one hundred and one"
       110187 "one hundred and ten thousand and one hundred and eighty seven"
       3122123 "three million and one hundred and twenty two thousand and one hundred and twenty three"
       68313373 "sixty eight million and three hundred and thirteen thousand and three hundred and seventy three"))

   (testing "the 'zero' edge case"
     (is (match? "zero" (parse/number->words client 0))))))


test/number_to_words/utils_test.clj
(ns number-to-words.utils-test
  (:require [number-to-words.utils :as utils]
            [matcher-combinators.test]
            [matcher-combinators.matchers :as m]
            [clojure.test :refer [deftest testing is]]))

(defn create-case [in out]
  {:in in :out out})

(deftest number->digit-groups
  (testing "should split number into trios of digit"
    (let [test-cases [{:in  1
                       :out (list (list 0 0 0) (list 0 0 0) (list 0 0 1))}
                      {:in  11
                       :out (list (list 0 0 0) (list 0 0 0) (list 0 1 1))}
                      {:in  111
                       :out (list (list 0 0 0) (list 0 0 0) (list 1 1 1))}
                      {:in  111111
                       :out (list (list 0 0 0) (list 1 1 1) (list 1 1 1))}
                      {:in  111111111
                       :out (list (list 1 1 1) (list 1 1 1) (list 1 1 1))}]]
      (is (match? (map :out test-cases)
                  (->> test-cases
                       (map :in)
                       (map utils/number->digit-groups)))))))

(deftest get-after-hundreds
  (testing "should get words for natural tens and units"
    (let [test-cases [(create-case [0 1] "one")
                      (create-case [0 2] "two")
                      (create-case [0 3] "three")
                      (create-case [0 4] "four")
                      (create-case [0 5] "five")
                      (create-case [0 6] "six")
                      (create-case [0 7] "seven")
                      (create-case [0 8] "eight")
                      (create-case [0 9] "nine")
                      (create-case [1 0] "ten")
                      (create-case [1 1] "eleven")
                      (create-case [1 2] "twelve")
                      (create-case [1 3] "thirteen")
                      (create-case [1 4] "fourteen")
                      (create-case [1 5] "fifteen")
                      (create-case [1 6] "sixteen")
                      (create-case [1 7] "seventeen")
                      (create-case [1 8] "eighteen")
                      (create-case [1 9] "nineteen")
                      (create-case [2 0] "twenty")
                      (create-case [3 0] "thirty")
                      (create-case [4 0] "forty")
                      (create-case [5 0] "fifty")
                      (create-case [6 0] "sixty")
                      (create-case [7 0] "seventy")
                      (create-case [8 0] "eighty")
                      (create-case [9 0] "ninety")]]
      (is (match? (map :out test-cases)
                  (->> test-cases
                       (map :in)
                       (map (fn [[tens units]] (prn (utils/get-after-hundreds tens units)) (utils/get-after-hundreds tens units)))))))
    (testing "should get words for non natural tens and units groups"
      (let [test-cases [(create-case [2 1] "twenty one")
                        (create-case [3 2] "thirty two")
                        (create-case [4 3] "forty three")
                        (create-case [5 4] "fifty four")
                        (create-case [6 5] "sixty five")
                        (create-case [7 6] "seventy six")
                        (create-case [8 7] "eighty seven")
                        (create-case [9 8] "ninety eight")
                        (create-case [9 9] "ninety nine")]]
      (is (match? (map :out test-cases)
                  (->> test-cases
                       (map :in)
                       (map (fn [[tens units]] (utils/get-after-hundreds tens units))))))))
    (testing "should get words for non natural tens and units groups"
      (let [create-case (fn [digits out]
                          {:in digits :out out})
            test-cases  [(create-case [0 0] nil)]]
      (is (match? (map :out test-cases)
                  (->> test-cases
                       (map :in)
                       (map (fn [[tens units]] (utils/get-after-hundreds tens units))))))))))

(deftest get-hundreds
    (testing "Should return the name without 'and' sentence"
      (let [test-cases [(create-case [1 0 0] "one hundred")
                        (create-case [2 0 0] "two hundred")
                        (create-case [3 0 0] "three hundred")
                        (create-case [4 0 0] "four hundred")
                        (create-case [5 0 0] "five hundred")
                        (create-case [6 0 0] "six hundred")
                        (create-case [7 0 0] "seven hundred")
                        (create-case [8 0 0] "eight hundred")
                        (create-case [9 0 0] "nine hundred")]]
      (is (match? (map :out test-cases)
                  (->> test-cases
                       (map :in)
                       (map (fn [in] (apply utils/get-hundreds in)))))))
    (testing "Should return the word with 'and' sentence"
      (let [test-cases [(create-case [1 0 1] "one hundred and")
                        (create-case [2 1 0] "two hundred and")
                        (create-case [3 1 1] "three hundred and")
                        (create-case [4 1 0] "four hundred and")
                        (create-case [5 0 1] "five hundred and")
                        (create-case [6 1 0] "six hundred and")
                        (create-case [7 1 1] "seven hundred and")
                        (create-case [8 1 1] "eight hundred and")
                        (create-case [9 0 1] "nine hundred and")]]
        (is (match? (map :out test-cases)
                    (->> test-cases
                         (map :in)
                         (map (fn [in] (apply utils/get-hundreds in))))))))))

(deftest group->word
  (testing "Should return the words with power specs"
    (let [test-cases [(create-case [0 [0 0 1]] "one million")
                      (create-case [1 [0 0 1]] "one thousand")
                      (create-case [1 [0 1 1]] "eleven thousand")
                      (create-case [2 [0 0 1]] "one")
                      (create-case [0 [1 3 8]] "one hundred and thirty eight million")
                      (create-case [1 [7 1 9]] "seven hundred and nineteen thousand")
                      (create-case [2 [3 4 7]] "three hundred and forty seven")]]
      (is (match? (map :out test-cases)
                    (->> test-cases
                         (map :in)
                         (map (fn [in] (apply utils/group->words in)))))))
    )
  (testing "Should return nil when every number is zero"
    (let [test-cases [(create-case [0 [0 0 0]] nil)
                      (create-case [1 [0 0 0]] nil)
                      (create-case [2 [0 0 0]] nil)]]
      (is (match? (map :out test-cases)
                  (->> test-cases
                       (map :in)
                       (map (fn [in] (apply utils/group->words in)))))))))


project.clj
(defproject io.github.yuhribernardes/number-to-words "1.0.0"
  :description "Library to convert number to words"
  :url "https://github.com/yuhribernardes/number-to-word"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[org.clojure/clojure "1.10.3"]]
  :main number-to-words.parse
  :target-path "target/%s"
  :aliases {"kaocha" ["with-profile" "+kaocha" "run" "-m" "kaocha.runner"]}
  :profiles {:uberjar {:aot :all
                       :jvm-opts ["-Dclojure.compiler.direct-linking=true"]}
             :kaocha {:dependencies [[lambdaisland/kaocha "1.62.993"]
                                     [nubank/matcher-combinators "3.3.1"]]}}
  :repositories {"release" {:url "https://repo.clojars.org/"
                            :username :env/clojars_user
                            :password :env/clojars_pwd}})


src/number_to_words/parse.clj
(ns number-to-words.parse
  (:require [number-to-words.client :as n-cli]
            [clojure.string :as str]))

(defn number->digit-groups [number]
  (->> (format "%09d" number)
       (partition 3 3)
       (map (fn [group] (map #(Character/getNumericValue %) group)))))

(defn get-after-hundreds [client group]
  (when-not (every? zero? group)
    (let [[tens units] group
          raw-ten (* tens 10)]
      (if (= tens 1)
        (->> (+ raw-ten units)
             (n-cli/get-natural-name client))
        (->> [raw-ten units]
             (map #(n-cli/get-natural-name client %1))
             (filter (comp not nil?))
             (str/join " "))))))

(defn get-hundreds [client [hundreds tens units]]
  (when-not (zero? hundreds)
    (let [and-sentence (when (not (every? zero? [tens units])) "and")
          hundreds-word (n-cli/get-natural-name client hundreds)]
      (->> [hundreds-word "hundred" and-sentence]
           (filter (comp not nil?))
           (str/join " " )))))

(defn group->words [client power group]
  (when-not (every? zero? group)
    (let [[_ tens units]   group
          power-complement (n-cli/get-powers-of-tens client power)]
      (->> [(get-hundreds client group)
            (get-after-hundreds client [tens units])
            power-complement]
           (filter (comp not nil?))
           (str/join " ")))))

(defn number->words
  "Take a number 'n' and convert it into words"
  [client n]
  (if (zero? n)
    (n-cli/get-zero-case client)
    (->> n
         number->digit-groups
         (map-indexed #(group->words client %1 %2))
         (filter (comp not empty?))
         (str/join " and "))))


src/number_to_words/specs.clj
(ns number-to-words.specs)

(def number-specs {:natural {1 "one"
                             2 "two"
                             3 "three"
                             4 "four"
                             5 "five"
                             6 "six"
                             7 "seven"
                             8 "eight"
                             9 "nine"
                             10 "ten"
                             11 "eleven"
                             12 "twelve"
                             13 "thirteen"
                             14 "fourteen"
                             15 "fifteen"
                             16 "sixteen"
                             17 "seventeen"
                             18 "eighteen"
                             19 "nineteen"
                             20 "twenty"
                             30 "thirty"
                             40 "forty"
                             50 "fifty"
                             60 "sixty"
                             70 "seventy"
                             80 "eighty"
                             90 "ninety"}
                   :powers-of-ten ["million" "thousand" nil]})

(defn get-natural-name [n]
  (get-in number-specs [:natural n]))

(defn get-powers-of-tens [n]
  (-> number-specs
      :powers-of-ten
      (nth n)))


src/number_to_words/utils.clj
(ns number-to-words.utils
  (:require [number-to-words.specs :as n-specs]
            [clojure.string :as str]))

(defn number->digit-groups [number]
  (->> (format "%09d" number)
       (partition 3 3)
       (map (fn [group] (map #(Character/getNumericValue %) group)))))

(defn get-after-hundreds [tens units]
  (when-not (and (zero? tens) (zero? units))
    (let [raw-ten (* tens 10)]
      (if (= tens 1)
        (do
          (->> (+ raw-ten units)
               n-specs/get-natural-name
               #_((fn [v] (prn v) v))))
        (->> [raw-ten units]
             (map n-specs/get-natural-name)
             (filter (comp not nil?))
             (str/join " "))))))

(defn get-hundreds [hundreds tens units]
  (when-not (zero? hundreds)
    (let [and-sentence (when (or (> tens 0) (> units 0)) "and")
          hundreds-word (n-specs/get-natural-name hundreds)]
      (->> [hundreds-word "hundred" and-sentence]
           (filter (comp not nil?))
           (str/join " " )))))

(defn group->words [power group]
  (when-not (every? zero? group)
    (let [[_ tens units]   group
          power-complement (n-specs/get-powers-of-tens power)]
      (->> [(apply get-hundreds group)
            (get-after-hundreds tens units)
            power-complement]
           (filter (comp not nil?))
           (str/join " ")))))

(comment
  (letfn [(groups->words [groups]
            (map-indexed group->words groups))]
    (let [sample 123456000]
      (->> sample
           number->digit-groups
           groups->words
           (filter (comp not empty?))
           (str/join " and ")
           #_(str/trim)))))


src/number_to_words/client.clj
(ns number-to-words.client)

(def specs {:en
            {:zero-case "zero"
             :natural  {1  "one"
                        2  "two"
                        3  "three"
                        4  "four"
                        5  "five"
                        6  "six"
                        7  "seven"
                        8  "eight"
                        9  "nine"
                        10 "ten"
                        11 "eleven"
                        12 "twelve"
                        13 "thirteen"
                        14 "fourteen"
                        15 "fifteen"
                        16 "sixteen"
                        17 "seventeen"
                        18 "eighteen"
                        19 "nineteen"
                        20 "twenty"
                        30 "thirty"
                        40 "forty"
                        50 "fifty"
                        60 "sixty"
                        70 "seventy"
                        80 "eighty"
                        90 "ninety"}
             :ten-pows ["million" "thousand" nil]}})

(defn create
  ([specs lang] (lang specs))
  ([lang] (create specs lang)))


(defn assoc-lang [old-spec lang {:keys [zero-case natural pows]}]
  (assoc old-spec lang {:zero-case zero-case
                        :natural  natural
                        :ten-pows pows}))

(defn get-zero-case [client]
  (:zero-case client))

(defn get-natural-name [client n]
  (get-in client [:natural n]))

(defn get-powers-of-tens [client n]
  (-> client
      :ten-pows
      (nth n)))
